File: 2019/src/aoc2019/day1.clj
(ns aoc2019.day1
  (:require [clojure.string :as str]))

(defn get-input [f] (str/split (slurp f) #"\n"))

; part1
(defn get-fuel [mass] (- (int (/ mass 3)) 2))

; part2
(defn total-fuel [mass]
  (loop [fuel (get-fuel mass) fuel-acc 0]
    (if-not (pos? fuel) fuel-acc
      (recur (get-fuel fuel) (+ fuel-acc fuel)))))

(defn -main [& args]
  (let [module-masses (map #(Integer/parseInt %) (get-input "day1.txt"))
        fuels (map get-fuel module-masses)
        total-fuels (map total-fuel module-masses)]
    (println "Part1: " (reduce + fuels))
    (println "Part2: " (reduce + total-fuels))))

File: 2020/src/main/java/Day1.java
import java.util.*;
import java.util.stream.Collectors;

public class Day1 {

    public static void main(String... args) {
        List<Integer> ints = Utils.readLines("day1.txt")
                .map(Integer::parseInt)
                .collect(Collectors.toList());

        int size = ints.size();

        for (int i = 0; i < size; i++) {
            for (int j = i; j < size; j++) {
                int a = ints.get(i);
                int b = ints.get(j);
                if (a + b == 2020) {
                    Utils.println("Day1 Part1: %d * %d == %d", a, b, a * b);
                }
            }
        }

        for (int i = 0; i < size; i++) {
            for (int j = i; j < size; j++) {
                for (int k = j; k < size; k++) {
                    int a = ints.get(i);
                    int b = ints.get(j);
                    int c = ints.get(k);
                    if (a + b + c == 2020) {
                        Utils.println("Day1 Part2: %d * %d * %d == %d", a, b, c, a * b * c);
                    }
                }
            }
        }
    }
}
